Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,321 @@
+    ///   - buffer: The buffer to read from.
+    /// - Returns: The integer that was read, or nil if it was not possible to read it.
+    func readInteger<IntegerType: FixedWidthInteger>(
+        as: IntegerType.Type,
+        from buffer: inout ByteBuffer
+    ) -> IntegerType?
+
+    /// Write an integer to a buffer. Move the writer index to after the written integer.
+    /// - Parameters:
+    ///    - integer: The type of the integer to write.
+    ///    - to: The buffer to write to.
+    func writeInteger<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        to buffer: inout ByteBuffer
+    ) -> Int
+
+    /// When writing an integer using this strategy, how many bytes we should reserve.
+    /// If the actual bytes used by the write function is more or less than this, it will be necessary to shuffle bytes.
+    /// Therefore, an accurate prediction here will improve performance.
+    var reservedCapacityForInteger: Int { get }
+
+    /// Write an integer to a buffer. The writer index should be moved accordingly.
+    /// This function takes  a `reservedCapacity` parameter which is the number of bytes we already reserved for writing this integer.
+    /// If you use write more or less than this many bytes, we wll need to move bytes around to make that possible.
+    /// Therefore, you may decide to encode differently to use more bytes than you actually would need, if it is possible for you to do so.
+    /// That way, you waste the reserved bytes, but improve writing performance.
+    func writeIntegerWithReservedCapacity(
+        _ integer: Int,
+        reservedCapacity: Int,
+        to buffer: inout ByteBuffer
+    ) -> Int
+}
+
+extension NIOBinaryIntegerEncodingStrategy {
+    @inlinable
+    public var reservedCapacityForInteger: Int { 1 }
+
+    @inlinable
+    public func writeIntegerWithReservedCapacity<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        reservedCapacity: Int,
+        to buffer: inout ByteBuffer
+    ) -> Int {
+        self.writeInteger(integer, to: &buffer)
+    }
+}
+
+extension ByteBuffer {
+    /// Read a binary encoded integer, moving the `readerIndex` appropriately.
+    /// If there are not enough bytes, nil is returned.
+    @inlinable
+    public mutating func readEncodedInteger<Strategy: NIOBinaryIntegerEncodingStrategy>(_ strategy: Strategy) -> Int? {
+        strategy.readInteger(as: Int.self, from: &self)
+    }
+
+    /// Write a binary encoded integer.
+    ///
+    /// - Returns: The number of bytes written.
+    @discardableResult
+    @inlinable
+    public mutating func writeEncodedInteger<
+        Integer: FixedWidthInteger,
+        Strategy: NIOBinaryIntegerEncodingStrategy
+    >(
+        _ value: Integer,
+        strategy: Strategy
+    ) -> Int {
+        strategy.writeInteger(value, to: &self)
+    }
+
+    /// Prefixes a message written by `writeMessage` with the number of bytes written using `strategy`.
```

Comment:
naming nit: this isn't necessarily a 'message' being written, 'bytes' is a little more generic